tests/gen/gen_auto.go

```diff
@@ -0,0 +1,242 @@
+	genConfigs, err := parseConfig(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := newRandomDocGenerator(typeDefs, genConfigs)
+	return generator.generateDocs(options...)
+}
+
+// AutoGenerate generates random documents from collection definitions.
+func AutoGenerate(definitions []client.CollectionDefinition, options ...Option) ([]GeneratedDoc, error) {
+	err := validateDefinitions(definitions)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, def := range definitions {
+		typeDefs[def.Description.Name] = def
+	}
+	generator := newRandomDocGenerator(typeDefs, nil)
+	return generator.generateDocs(options...)
+}
+
+func newRandomDocGenerator(types map[string]client.CollectionDefinition, config configsMap) *randomDocGenerator {
+	if config == nil {
+		config = make(configsMap)
+	}
+	configurator := newDocGenConfigurator(types, config)
+	return &randomDocGenerator{
+		configurator:  configurator,
+		generatedDocs: make(map[string][]genDoc),
+	}
+}
+
+type genDoc struct {
+	docKey string
+	doc    *client.Document
+}
+
+type randomDocGenerator struct {
+	configurator docsGenConfigurator
+
+	generatedDocs map[string][]genDoc
+	random        rand.Rand
+}
+
+func (g *randomDocGenerator) generateDocs(options ...Option) ([]GeneratedDoc, error) {
+	err := g.configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *g.configurator.random
+
+	resultDocs := make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	err = g.generateRandomDocs(g.configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, colName := range g.configurator.typesOrder {
+		typeDef := g.configurator.types[colName]
+		for _, doc := range g.generatedDocs[colName] {
+			resultDocs = append(resultDocs, GeneratedDoc{
+				Col: &typeDef,
+				Doc: doc.doc,
+			})
+		}
+	}
+	return resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.configurator.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.configurator.usageCounter.getNextTypeIndForField(secondaryType, field)
+	return g.generatedDocs[field.Schema][ind].docKey
+}
+
+func (g *randomDocGenerator) generateRandomDocs(order []string) error {
+	for _, typeName := range order {
+		col := DocsList{ColName: typeName}
+		typeDef := g.configurator.types[typeName]
+
+		currentTypeDemand := g.configurator.docsDemand[typeName]
+		averageDemand := currentTypeDemand.getAverage()
+		// we need to decide how many documents to generate for this type
+		// and if it's a range (say, 10-30) we take average (20).
```

Comment:
question: Why would anyone specify a range, if an average is always used?

Reply:
```graphql
type User {
    name: String
    devices: [Device] # max: 1, max 5
}
...
```
If you demand 100 `User` to generate, then the demand for `Device` is between 100 and 500, which is range.

Reply:
The range is never used though, in the above example there would never be any users with 1 or 5 devices, every user would have 3 devices no?

Reply:
no, that's not how it works. In the given example the average would be 300, which means that `Device` docs will be generated. These 300 devices will be randomly distributed across 100 users. So yes, each user will have from 1 to 5  devices randomly.

Reply:
Ah okay nice! Thanks :)

Where does that take place though?  This loop would always be `3` though, does this mean that every `Device` would be linked to exactly `3` Users?  Do you validate somewhere that the random stuff does not pick the same 3 Users for every `Device`?

Reply:
> Where does that take place though? 

> Do you validate somewhere else that the random stuff does not pick the same 3 Users for every Device?

you wanna take a look `(relationUsage) useNextDocKey()`

> This loop would always be 3 though, does this mean that every Device would be linked to exactly 3 Users?

With the example above this loop will make 300 iterations to produce 300 devices. Users will be randomly picked from the pool of available users for this particular relation.

Reply:
> With the example above this loop will make 300 iterations to produce 300 devices. Users will be randomly picked from the pool of available users for this particular relation.

In that example would the input param `order` have `100` "User" strings in it?

Reply:
> In that example would the input param order have 100 "User" strings in it?

`order` is the ordered list of collection names that need to be generated from left to right `[]string{"User", "Device", "Component"}`

Reply:
In this User-Device example `averageDemand` would equal `3` though no?  And each iteration of this loop appears to only add a single document, where is this multiplied by 100?

Reply:
I think you need to read the conversation again :) I already mentioned it:

> In the given example the average would be 300, which means that Device docs will be generated.

> With the example above this loop will make 300 iterations to produce 300 devices

`demand` is total demand of documents of a collection. It's not demand of subdocuments per document.